### PR TITLE
Handle feature references another sequence

### PR DESCRIPTION
**Pull request by peterjc**
This pull request addresses issue #808, including the work of Adam Sjøgren via the mailing list.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Rebasing and dealing with conflicts and new release...

**Comment by peterjc** [comment]
I haven't forgotten about this, but we've been distracted with the Python 2/3 cleanup in the meantime.

**Comment by peterjc** [comment]
Rebasing again (and leaving out the dual licensing changes)...

**Comment by peterjc** [comment]
Code seems to be working fine, but have not found a good example yet - for instance none of the features here give the expected translation:

```python
from Bio import SeqIO

files = ("FP885871.gbk", "FP885876.gbk")

references = SeqIO.to_dict(SeqIO.read(_, "gb") for _ in files)

for acc, ref in references.items():
    print(f"{acc} length {len(ref)} has {len(ref.features)} features")
    for f in ref.features:
        if f.type == "CDS" and len(f.location.parts) > 1:
            print(f"{f.type} feature at {f.location}")
            #print(f)
            expected = f.qualifiers["translation"][0]
            try:
                translation = f.extract(ref.seq, references).translate(cds=True)
            except:
                translation = f.extract(ref.seq, references).translate(to_stop=True)
            print(expected + "<-- expected")
            print(translation + "<-- computed", expected == translation)
            if "exception" in f.qualifiers and f.qualifiers["exception"] == ["RNA editing"]:
                continue
            assert expected == translation, f"{expected} vs {translation}"
```

Some (all?) of them have human readable comments explaining why (in more detail than just "RNA editing") at least.

**Comment by peterjc** [comment]
Sadly this example from ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Arabidopsis_thaliana/ used in the Tutorial from drawing tRNA on chromosomes does not appear to have any clear trans-splicing annotation:

```python
filenames = [
    "CHR_I/NC_003070.gbk",
    "CHR_II/NC_003071.gbk",
    "CHR_III/NC_003074.gbk",
    "CHR_IV/NC_003075.gbk",
    "CHR_V/NC_003076.gbk",
]

references = SeqIO.index_db("Arabidopsis.idx", filenames, "genbank")
```

Perhaps a newer version might... or maybe this is just not an appropriate example.

**Comment by peterjc** [comment]
Zebrafish to the rescue - a well behaved example from Adam Sjøgren, BX465837.11 which has a couple of CDS with trans-splicing with BX537337.9

https://www.ncbi.nlm.nih.gov/nuccore/BX465837.11
https://www.ncbi.nlm.nih.gov/nuccore/BX537337.9

Should work in EMBL format too:

https://www.ebi.ac.uk/ena/browser/api/embl/BX465837.11
https://www.ebi.ac.uk/ena/browser/api/embl/BX537337.9

**Comment by peterjc** [comment]
@MarkusPiotrowski - would you be a good person to review this?

I'm hoping to turn the Zebrafish example into something for the Tutorial and/or docstrings.

**Comment by peterjc** [comment]
This is going to need some merge conflicts resolved now...

**Comment by peterjc** [comment]
@mdehoon do you want to review?

**Comment by mdehoon** [comment]
I guess these exceptions:
```
            raise ValueError("Feature references another sequence.")
```
in `_shift` and `_flip` can now be removed?

**Comment by peterjc** [comment]
Yes, as per my comment https://github.com/biopython/biopython/issues/808#issuecomment-683291626 in those cases ``_shift`` and ``_flip`` are relative to the local sequence, and thus are a no-operation relative to an external sequence.

File: Bio/SeqFeature.py [hunk 1]
```diff
@@ -1004,22 +1006,26 @@ def __eq__(self, other):
...
     def _flip(self, length):
-        """Return a copy of the location after the parent is reversed (PRIVATE)."""
+        """Return a copy of the location after the parent is reversed (PRIVATE).
+
+        Returns self when location is relative to an external reference.
+        """
         if self.ref or self.ref_db:
-            # TODO - Return self?
-            raise ValueError("Feature references another sequence.")
+            self
```
Comment (mdehoon) [inline Bio/SeqFeature.py:1028]: return self

Comment (peterjc) [reply Bio/SeqFeature.py:1028]: Thank you. I'm slightly surprised flake8 didn't complain about that.

**Review by mdehoon — APPROVED** [review]

**Comment by peterjc** [comment]
Thank you.

**Pull request merged**